torch/csrc/jit/runtime/profiling_record.h

```diff
@@ -8,12 +8,127 @@
+//     * Partitioning can be done even before we have seen all the dimension locations associated with `ShapeSymbol S1` 
+//       * We use `getSymbolInSet` of `ShapeSymbolTable` to remember all  `ShapeSymbols` from run #2 we observed in the dimension locations associated with `ShapeSymbol S1` .
+//       * For every `ShapeSymbol` from run #2 in the dimension location associated with `ShapeSymbol S1`  `getSymbolInSet` returns a symbol that we assign to the dimension location in a new TensorType.
+//         * It's important to point out that the same `ShapeSymbol S2` from run #2 in two dimension locations that have different `ShapeSymbol`s in run #1 are different! These dimension locations will belong to different subsets and have different `ShapeSymbol`s after merge.
+//         * On the other hand, for the same `ShapeSymbol S2` in two dimension locations that have `ShapeSymbol S1` in run #1`getSymbolInSet` will return the same symbol.  
+
 namespace torch {
 namespace jit {
 
 using ::c10::TensorTypePtr;
+using Dimension = int64_t;
+
+struct ProfilingRecord;
+
+// `SetPartitioningHelper` is used to maintain the following invariant:
+// For **every** profiling run, *the same `ShapeSymbol` is always associated 
+// with the same `Dimension`*.
+// while merging the profiling information from multiple runs.
+struct SetPartitioningHelper {
+  
+  std::map<c10::ShapeSymbol, std::map<Dimension, c10::ShapeSymbol>> sets2subsets_;
+
+  // `partitionSetByDimension` partitions a virtual set  
+  // of dimension locations associated with ShapeSymbol `symbol` into subsets.
+  // Partitioning is equivalent to giving (or renaming) a particular  
+  // dimension location a new `ShapeSymbol`.
+  // The same `Dimension` value in different dimension locations
+  // that used to have `symbol` will receive the same
+  // new `ShapeSymbol`, effectively forming a new set.
+  c10::ShapeSymbol partitionSetByDimension(
+    Dimension new_size,
+    c10::ShapeSymbol symbol) {
+    auto& dims2symbols = getSetForSymbol(symbol);
+
+    if (dims2symbols.count(new_size) == 0) {
+      auto new_sym = c10::ShapeSymbol::newSymbol();
+      dims2symbols[new_size] = new_sym;
+      return new_sym;
+    }
+
+    return dims2symbols[new_size];
+  }
+
+  private:
+
+    std::map<Dimension, c10::ShapeSymbol>& getSetForSymbol(c10::ShapeSymbol s) {
+      auto& set = sets2subsets_[s];
+      // N.B. adding a mapping { s.static_size(), s }
+      // makes sure we preserve the fact that
+      // some dimension values remain the same
+      // across all profiled runs
+      if (s.is_static()) {
+        set.insert({s.static_size(), s});
+      }
+      return set;
+    }
+};
+
+// ShapeSymbolTable is used by Interpreter
+// to assign dimension values to ShapeSymbols
+// and fail a guard if the same symbol
+// is assigned more than one dimension value.
+struct ShapeSymbolTable {
```

Comment:
A class used by the interpreter but not specific to profiling should probably go in a different file.